File: samples/android-app/src/main/kotlin/io/github/andrewmalitchuk/uncial/samples/android/ui/source/kit/atom/chip/Chip.kt
package io.github.andrewmalitchuk.uncial.samples.android.ui.source.kit.atom.chip

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.padding
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.unit.dp
import io.github.andrewmalitchuk.uncial.samples.android.ui.core.Theme

/**
 * A small piece of state with a box around it.
 *
 * [ChipTone.Unavailable] exists because the whole point of the SDK's capability reporting
 * is that platforms differ; a UI that hides what it cannot do is the UI this sample is
 * arguing against.
 */
@Composable
fun Chip(
    text: String,
    modifier: Modifier = Modifier,
    tone: ChipTone = ChipTone.Neutral,
) {
    val color = Theme.color
    val shape = Theme.corner.control

    val background = when (tone) {
        ChipTone.Neutral -> color.surface
        ChipTone.Selected -> color.selected
        ChipTone.Accent -> color.accentMuted
        ChipTone.Unavailable -> color.surfaceVariant
    }
    val foreground = when (tone) {
        ChipTone.Neutral, ChipTone.Selected -> color.ink
        ChipTone.Accent -> color.accent
        ChipTone.Unavailable -> color.inkSubtle
    }
    val outline = when (tone) {
        ChipTone.Neutral -> color.outline
        else -> Color.Transparent
    }

    Text(
        text = text,
        style = Theme.typography.caption,
        color = foreground,
        modifier = modifier
            .clip(shape)
            .background(background, shape)
            .border(Theme.size.hairline, outline, shape)
            .padding(horizontal = 10.dp, vertical = 5.dp),
    )
}

File: samples/ios-app/UncialSample/Kit/Chip.swift
import SwiftUI

/// What a ``Chip`` is saying about the thing it is attached to.
enum ChipTone {
    /// A fact: a language, a code, a count.
    case neutral
    /// A fact the user chose, or one the engine confirmed.
    case selected
    /// Something the engine is doing — the OCR path, an active engine.
    case accent
    /// Something this device cannot do. Shown, never hidden.
    case unavailable
}

/// A small piece of state with a box around it.
///
/// ``ChipTone/unavailable`` exists because the whole point of the SDK's capability
/// reporting is that platforms differ; a UI that hides what it cannot do is the UI this
/// sample argues against.
struct Chip: View {

    let text: String
    var tone: ChipTone = .neutral

    var body: some View {
        Text(text)
            .font(Theme.typography.caption)
            .foregroundColor(foreground)
            .padding(.horizontal, 10)
            .padding(.vertical, 5)
            .background(background)
            .clipShape(Capsule())
            .overlay(Capsule().stroke(border, lineWidth: Theme.size.hairline))
    }

    private var background: Color {
        switch tone {
        case .neutral: return Theme.color.surface
        case .selected: return Theme.color.selected
        case .accent: return Theme.color.accentMuted
        case .unavailable: return Theme.color.surfaceVariant
        }
    }

    private var foreground: Color {
        switch tone {
        case .neutral, .selected: return Theme.color.ink
        case .accent: return Theme.color.accent
        case .unavailable: return Theme.color.inkSubtle
        }
    }

    private var border: Color {
        tone == .neutral ? Theme.color.outline : .clear
    }
}
